### PR TITLE
log in and log out

### PR DESCRIPTION
**Pull request by thanhnga67**

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    if user && user.authenticate(params[:session][:password])
+      log_in user
+      params[:session][:remember_me] == "1" ? remember(user) : forget(user)
+      remember user
+      redirect_to user
+    else
+      flash.now[:danger] = t("invaild")
+      render :new
+    end
+  end
```
Comment (tunv) [inline app/controllers/sessions_controller.rb]: add white line

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
+class SessionsController < ApplicationController
+  def new
+  end
+
+  def create
+    user = User.find_by(email: params[:session][:email].downcase)
+    if user && user.authenticate(params[:session][:password])
+      log_in user
+      params[:session][:remember_me] == "1" ? remember(user) : forget(user)
```
Comment (leminhtuan2015) [inline app/controllers/sessions_controller.rb]: Place 1 to Contanst of Setting

File: app/controllers/sessions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+      remember user
+      redirect_to user
+    else
+      flash.now[:danger] = t("invaild")
```
Comment (leminhtuan2015) [inline app/controllers/sessions_controller.rb]: t ""

File: app/models/user.rb [hunk 4]
```diff
@@ -9,4 +10,28 @@ class User < ActiveRecord::Base
   validates :email, presence: true, format: {with: FORMAT}, uniqueness:{case_sensitive: false}
   validates :password, presence: true, length: {minimum: Settings.user.minimum}
   validates :password_confirmation, presence: true
+
+  def User.digest string
+    cost = ActiveModel::SecurePassword.min_cost ? BCrypt::Engine::MIN_COST : 
+    BCrypt::Engine.cost
+    BCrypt::Password.create(string, cost: cost)
```
Comment (leminhtuan2015) [inline app/models/user.rb]: ( )

File: app/models/user.rb [hunk 5]
```diff
@@ -9,4 +10,28 @@ class User < ActiveRecord::Base
...
+  end
+  
+  def remember
+    self.remember_token = User.new_token
+    update_attribute(:remember_digest, User.digest(remember_token))
+  end
+
+  def authenticated? remember_token
+  	return false if remember_digest.nil?
+    BCrypt::Password.new(remember_digest).is_password?(remember_token)
```
Comment (leminhtuan2015) [inline app/models/user.rb]: ( )

File: app/models/user.rb [hunk 6]
```diff
@@ -9,4 +10,28 @@ class User < ActiveRecord::Base
...
+  end
+
+  def forget
+    update_attribute(:remember_digest, nil)
```
Comment (leminhtuan2015) [inline app/models/user.rb]: ( )

File: app/views/layouts/_header.html.erb [hunk 7]
```diff
@@ -1,11 +1,28 @@
...
- <div class="container">
+  <div class="container">
     <%= link_to t("title"), root_path, id: "logo" %>
     <nav>
       <ul class="nav navbar-nav navbar-right">
         <li><%= link_to t("home"), root_path %></li>
         <li><%= link_to t("help"), help_path %></li>
-        <li><%= link_to t("login"), "#" %></li>
+        <% if logged_in? %>
+          <li><%= link_to "Users", "#" %></li>
```
Comment (leminhtuan2015) [inline app/views/layouts/_header.html.erb]: I18n

File: app/views/layouts/_header.html.erb [hunk 8]
```diff
@@ -1,11 +1,28 @@
...
+          <li class="dropdown">
+            <a href="#" class="dropdown-toggle" data-toggle="dropdown">
+              Account <b class="caret"></b>
```
Comment (leminhtuan2015) [inline app/views/layouts/_header.html.erb]: link_to

File: app/views/sessions/_form.html.erb [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+  <%= f.label :remember_me, class: "checkbox inline" do %>
+    <%= f.check_box :remember_me %>
+    <span><%= t("remember_me") %></span>
+  <% end %>
+
+  <%= f.submit t("login"), class: "btn btn-lg btn btn-success" %>
+<% end %>
+
+<p>New user? <%= link_to t("signup"), signup_path %></p>
```
Comment (leminhtuan2015) [inline app/views/sessions/_form.html.erb]: I18n

File: app/views/sessions/_form.html.erb [hunk 10]
```diff
@@ -0,0 +1,17 @@
+<%= form_for(:session, url: login_path) do |f| %>
```
Comment (leminhtuan2015) [inline app/views/sessions/_form.html.erb]: ( )

File: app/views/sessions/_form.html.erb [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+
+  <%= f.label :email %>
+  <%= f.email_field :email, class: 'form-control' %>
+
+  <%= f.label :password %>
+  <%= f.password_field :password, class: 'form-control' %>
```
Comment (leminhtuan2015) [inline app/views/sessions/_form.html.erb]: '

File: app/views/sessions/_form.html.erb [hunk 12]
```diff
@@ -0,0 +1,17 @@
+<%= form_for(:session, url: login_path) do |f| %>
+
+  <%= f.label :email %>
+  <%= f.email_field :email, class: 'form-control' %>
```
Comment (leminhtuan2015) [inline app/views/sessions/_form.html.erb]: '

File: app/views/sessions/_form.html.erb [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+
+  <%= f.label :password %>
+  <%= f.password_field :password, class: 'form-control' %>
+
+  <%= f.label :remember_me, class: "checkbox inline" do %>
+    <%= f.check_box :remember_me %>
+    <span><%= t("remember_me") %></span>
```
Comment (leminhtuan2015) [inline app/views/sessions/_form.html.erb]: ( )

File: db/seeds.rb [hunk 14]
```diff
@@ -1,7 +1,9 @@
-# This file should contain all the record creation needed to seed the database with its default values.
-# The data can then be loaded with the rake db:seed (or created alongside the db with db:setup).
-#
-# Examples:
-#
-#   cities = City.create([{ name: 'Chicago' }, { name: 'Copenhagen' }])
-#   Mayor.create(name: 'Emanuel', city: cities.first)
+User.create!(name:  "Example User", email: "example@railstutorial.org",
```
Comment (leminhtuan2015) [inline db/seeds.rb]: Space

File: db/seeds.rb [hunk 15]
```diff
@@ -1,7 +1,9 @@
...
+  password: "123456", password_confirmation: "123456")
+
+99.times do |n|
+  name  = Faker::Name.name
+  email = "example-#{n+1}@railstutorial.org"
+  password = "password"
+  User.create!(name:  name, email: email, password: password, password_confirmation: password)
```
Comment (leminhtuan2015) [inline db/seeds.rb]: Space

File: app/controllers/sessions_controller.rb [hunk 16]
```diff
@@ -0,0 +1,22 @@
+class SessionsController < ApplicationController
+  def new
+  end
+
+  def create
+    user = User.find_by(email: params[:session][:email].downcase)
```
Comment (VoTaiTri) [inline app/controllers/sessions_controller.rb]: try 
```ruby
User.find_by email: params[:session][:email].downcase
```

File: app/models/user.rb [hunk 17]
```diff
@@ -9,4 +10,28 @@ class User < ActiveRecord::Base
...
+    BCrypt::Password.create string, cost: cost
+  end
+
+  def User.new_token
+    SecureRandom.urlsafe_base64
+  end
+  
+  def remember
+    self.remember_token = User.new_token
+    update_attribute(:remember_digest, User.digest(remember_token))
```
Comment (VoTaiTri) [inline app/models/user.rb]: try
```ruby
update_attribute :remember_digest, User.digest(remember_token)

File: app/models/user.rb [hunk 18]
```diff
@@ -9,4 +10,28 @@ class User < ActiveRecord::Base
...
+  end
+
+  def authenticated? remember_token
+  	return false if remember_digest.nil?
```
Comment (VoTaiTri) [inline app/models/user.rb]: indent

File: app/views/sessions/new.html.erb [hunk 19]
```diff
@@ -0,0 +1,8 @@
+<% provide(:title, t("login")) %>
```
Comment (VoTaiTri) [inline app/views/sessions/new.html.erb]: try
```ruby
provide :title, t("login")

File: config/routes.rb [hunk 20]
```diff
@@ -1,10 +1,11 @@
...
-  
   get "help" => "static_pages#help"
   get "signup" => "users#new"
-  get "edit" => "users#edit"
+  get "login" => "sessions#new"
+  post "login" => "sessions#create"
+  delete "logout" => "sessions#destroy"
 
-  resources :users, except: [:index, :destroy]
+  resources :users
```
Comment (VoTaiTri) [inline config/routes.rb]: why??? 
don't change file user_controller in this pull

File: app/models/user.rb [hunk 21]
```diff
@@ -9,4 +10,28 @@ class User < ActiveRecord::Base
...
+    BCrypt::Password.create string, cost: cost
+  end
+
+  def User.new_token
+    SecureRandom.urlsafe_base64
+  end
+  
+  def remember
+    self.remember_token = User.new_token
+    update_attribute :remember_digest, User.digest(remember_token)
```
Comment (dieunb) [inline app/models/user.rb]: I want to you try `update_attributes!` and then compare with `update_attribute`.

File: app/models/user.rb [hunk 22]
```diff
@@ -9,4 +10,28 @@ class User < ActiveRecord::Base
...
+  end
+
+  def authenticated? remember_token
+    return false if remember_digest.nil?
+    BCrypt::Password.new remember_digest.is_password? remember_token
+  end
+
+  def forget
+    update_attribute :remember_digest, nil
```
Comment (dieunb) [inline app/models/user.rb]: I want to you try `update_attributes!` and then compare with `update_attribute`.

**Pull request merged**